File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}


interface summaryFormat {
  name: string,
  cookTime: number,
  ingredients: requiredItem[]
};

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: cookbookEntry[] = [];

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  }
  res.json({ msg: parsed_string });
  return;
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that
const parse_handwriting = (recipeName: string): string | null => {
  recipeName =recipeName.replace(/[-_]/g, " ")
  recipeName = recipeName.replace(/ +/g, " ")
  recipeName =recipeName.replace(/^ /, "")
  recipeName =recipeName.replace(/ $/, "")
  recipeName = recipeName.replace(/[^a-zA-Z ]/g, "");
  let words = recipeName.split(" ")
  let capatilised = words.map((word:string) => {
    return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase()
  })
  recipeName = capatilised.join(" ")
  if (recipeName.length === 0) return null
  return recipeName
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook

const writeCookbook = (obj:recipe | ingredient) => {
  cookbook.push(obj)
}

// check if the cookbook contains an entry with a certain name
const checkContains = (name:string):boolean => {
  return (cookbook.filter((obj:cookbookEntry) => obj.name === name).length !== 0)
}

// check if an object contains a specific key
const hasKey = (obj:any, str:string) => {
  return str in obj;
}
// return true if obj:any is a valid recipe, false otherwise
const checkValidRecipe = (obj:any): boolean =>{
  if (!isCookbookEntry(obj)) return false
  if (!(obj.type === "recipe")) return false
  if(!hasKey(obj, "requiredItems")) return false;
  if (!Array.isArray(obj.requiredItems)) return false;
  let initialSize = obj.requiredItems.length;
  obj.requiredItems.filter((item:any) => {
    return hasKey(item, "quantity") && hasKey(item, "name") && (Object.keys(item).length === 2) && (item.quantity > 0)
  })
  if (initialSize !== obj.requiredItems.length) return false;
  return true;
}
// return true if obj:any is a valid ingredient, false otherwise
const checkValidIngredient = (obj:any):boolean => {
  if (!isCookbookEntry(obj)) return false
  if (!(obj.type === "ingredient")) return false
  if(!hasKey(obj, "cookTime")) return false;
  if (parseInt(obj.cookTime) < 0) return false;
  return true;
}

const isCookbookEntry = (obj:any) => {
  if (!hasKey(obj, "name")) return false;
  if (!hasKey(obj, "type")) return false;
  return true;
}
const validItem = (obj:any): boolean => {
  return (checkValidIngredient(obj) || checkValidRecipe(obj))
}

const convertToEntry = (obj:any): recipe | ingredient | null => {
  if (!validItem(obj)) return null;
  if (obj.type === "recipe") return obj as recipe
  else return obj as ingredient
}

app.post("/entry", (req:Request, res:Response) => {
  const body = req.body
  let cookObj = convertToEntry(body);
  if (checkContains(body.name)) {
    res.status(400).send("already in the cookbook")
  }
  else if (cookObj === null) {
    res.status(400).send("could not convert to recipe or ingredient")
  }
  else {
    writeCookbook(cookObj)
    res.status(200).send("entry has been logged")
  }
});
// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name

const getEntry = (name:string): cookbookEntry | null => {
  for (let item of cookbook) {
    if (item.name === name && item.type === "recipe") return item as recipe;
    else if (item.name === name && item.type === "ingredient") return item as ingredient;
  }
  return null
}
// returns -1 if it can't find one  of the items, otherwise it returns the cookTime
// map will contain a count of the quantity of each ingredient
const addIngredientsToMap = (item:cookbookEntry, map:Map<string, number>, amount:number):number => {
  if (item.type === "ingredient") {
    let ingr = item as ingredient;
    if (map.has(item.name)) {
      map.set(item.name, map.get(item.name) + amount)
    }
    else {
      map.set(item.name, amount);
    }
    return ingr.cookTime * amount;
  }
  else {
    let total = 0;
    let recipe:recipe = item as recipe;
    for (let ingr of recipe.requiredItems) {
      if (!checkContains(ingr.name)) return -1;
      let entry:cookbookEntry = getEntry(ingr.name)
      let increment = addIngredientsToMap(entry, map, ingr.quantity * amount)
      if (increment === -1) return -1;
      total += increment
    }
    return total
  }
}
app.get("/summary", (req:Request, res:Request) => {
  const headers = req.query;
  if(!hasKey(headers, "name")){
    res.status(400).send("incorrect name field")
    return;
  }
  const name = headers.name;
  if (!checkContains(name)) {
    res.status(400).send("does not contain an entry with this name")
    return;
  }
  const entry:cookbookEntry = getEntry(name)
  if (entry.type !== "recipe") {
    res.status(400).send("entry is not a recipe")
    return;
  }
  const recipe:recipe = entry as recipe;
  let result:summaryFormat = {
    name: recipe.name,
    cookTime: 0,
    ingredients: []
  }
  let map = new Map()
  let cookTime = addIngredientsToMap(recipe, map, 1)
  if (cookTime === -1) {
    res.status(400).send("required item not found in the cookbook")
    return;
  }
  result.cookTime = cookTime;
  map.forEach((value, key) => { // add each of the values in the map to the ingredients array within the result object
    let obj = {
      name:key,
      quantity: value
    }
    result.ingredients.push(obj as requiredItem)
  })
  res.status(200).send(result)
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int

# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = []

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = recipeName.lower()
	recipeName = re.sub(r'[-_]', ' ', recipeName)
	recipeName = re.sub(r'[^a-z ]', '', recipeName)
	recipeName = re.sub(r' +', ' ', recipeName)
	recipeName = re.sub(r'^ ', '', recipeName)
	recipeName = re.sub(r' $', '', recipeName)
	if len(recipeName) == 0:
		return None
	splitStr = recipeName.split(' ')
	recipeName = ''
	for word in splitStr:
		if len(word) > 1:
			recipeName += word[0].upper() + word[1:] + ' '
		else:
			recipeName += word[0].upper()
	recipeName = re.sub(r' $', '', recipeName) # remove trailing white space
	return recipeName


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook


# return true if the object data is a recipe object
def parseRecipe(data):
	if data['type'] != "recipe":
		return False
	if 'requiredItems' not in data:
		return False
	if not isinstance(data['requiredItems'], list):
		return False
	for item in data['requiredItems']:
		if 'name' not in item:
			return False
		if 'quantity' not in item:
			return False
		if len(item) != 2:
			return False
	return True

# return true if the object data is an ingredient object
def parseIngredient(data):
	if data['type'] != "ingredient":
		return False
	if 'cookTime' not in data:
		return False
	if int(data['cookTime']) < 0:
		return False
	return True

# return true if name is in the cookbook
def cookbookContains(name):
	for entry in cookbook:
		if entry['name'] == name:
			return True
	return False

# insert the object data into the cookbook, assumes that data is valid
def insertIntoCookbook(data):
	cookbook.append(data)

@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()
	if 'name' not in data:
		return 'key name not found', 400
	if 'type' not in data:
		return 'key type not found', 400
	if cookbookContains(data['name']):
		return 'entry already in cookbook', 400
	if parseRecipe(data):
		insertIntoCookbook(data)
		return 'recipe logged', 200
	elif parseIngredient(data):
		insertIntoCookbook(data)
		return 'ingredient logged', 200
	return 'invalid entry', 400


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name

def getEntry(name):
	for entry in cookbook:
		if entry['name'] == name:
			return entry
	return None


# returns the cooking time for entry 'name'
# updates mp (ingredientName -> quantity) with the quantity of each ingredient
# if an entry with name 'name' isn't in the cookbook return -1. if name is a recipe and one of the required items is
# invalid return -1
def getIngredients(name, mp, amount):
	entry = getEntry(name)
	if entry == None:
		return -1
	if entry['type'] == 'ingredient':
		if entry['name'] in mp:
			mp[entry['name']] += amount
		else:
			mp[entry['name']] = amount
		return amount * entry['cookTime']
	else:
		total = 0
		for item in entry['requiredItems']:
			entry = getEntry(item['name'])
			if entry == None:
				return -1
			cookTime = getIngredients(entry['name'], mp, amount * item['quantity'])
			if cookTime < 0:
				return -1
			total += cookTime
		return total

@app.route('/summary', methods=['GET'])
def summary():
	name = request.args.get('name')
	if not cookbookContains(name):
		return 'entry not found', 400
	entry = getEntry(name)
	if entry['type'] != 'recipe':
		return 'entry is not a recipe', 400
	mp = {}
	cookTime = getIngredients(entry['name'], mp, 1) # this loades the base ingredients into mp in the form (ingredientName -> quantity)
	if cookTime < 0:
		return 'required item not found in the cookbook', 400
	ingredients = []
	for key in mp:
		ingredients.append({'name': key, 'quantity': mp[key]})
	return {'name': entry['name'], 'cookTime': cookTime, 'ingredients': ingredients}, 200


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
